modules/mod-deep-learning/DeepModelManager.cpp

```diff
@@ -0,0 +1,554 @@
+{
+   ModelCardHolder card = std::make_shared<ModelCard>();
+   return card;
+}
+
+bool DeepModelManager::NewCardFromHuggingFace(ModelCardHolder card, const std::string &jsonBody, const std::string &repoID)
+{
+   wxStringTokenizer st(wxString(repoID), wxT("/"));
+   const std::string cardAuthor = audacity::ToUTF8(st.GetNextToken());
+   const std::string cardName = audacity::ToUTF8(st.GetNextToken());
+   
+   try
+   {
+      Doc doc = parsers::ParseString(jsonBody);
+      card->Deserialize(doc, mModelCardSchema);
+      (*card).name(cardName)
+             .author(cardAuthor)
+             .local(false)
+             .local_path(audacity::ToUTF8(GetRepoDir(card)));
+
+      return true;
+   }
+   catch (const InvalidModelCardDocument &e)
+   {
+      wxLogError(wxString(e.what()));
+      return false;
+   }
+   catch (const char *msg)
+   { 
+      wxLogError(wxString(msg));
+      wxASSERT(false);
+      return false;
+   }
+}
+
+bool DeepModelManager::NewCardFromLocal(ModelCardHolder card, const std::string &filePath)
+{
+   try
+   {
+      std::string localPath = audacity::ToUTF8(wxFileName(wxString(filePath)).GetPath());
+      card->DeserializeFromFile(filePath, mModelCardSchema);
+      (*card).local(true)
+             .local_path(localPath);
+      return true;
+   }
+   catch (const InvalidModelCardDocument &e)
+   {
+      return false;
+   }
+}
+
+network_manager::ResponsePtr DeepModelManager::DownloadModel
+(ModelCardHolder card, ProgressCallback onProgress, CompletionHandler onCompleted)
+{
+   using namespace network_manager;
+
+   // TODO: this is not raising an error for an invalid URL 
+   // try adding a double slash anyuwhere to break it. 
+   // its because huggingface returns 200s saying "Not Found"
+   const std::string url = GetFileURL(card->GetRepoID(), "model.pt");
+   
+   wxLogDebug("downloading from %s", url);
+
+   Request request(url);
+
+   // send request
+   NetworkManager &manager = NetworkManager::GetInstance();
+   ResponsePtr response = manager.doGet(request);
+
+   // open a file to write the model to 
+   const std::string repoId = card->GetRepoID(); 
+   wxString path = wxFileName(this->GetRepoDir(card), "model.pt").GetFullPath();
+   std::shared_ptr<wxFile> file = std::make_shared<wxFile>(path, wxFile::write);
+
+   // set callback for download progress
+   if (onProgress)
+      response->setDownloadProgressCallback(onProgress);
+
+   // completion handler
+   response->setRequestFinishedCallback(
+      [response, handler = std::move(onCompleted)](IResponse*) 
+      {
+         const std::string responseData = response->readAll<std::string>();
+
+         if (handler)
+            handler(response->getHTTPCode(), responseData);
+      }
+   );
+
+   // write to file here
+   response->setOnDataReceivedCallback(
+      [this, response, card, file](IResponse*) 
+      {
+         // abort
+         if (!this->IsInstalling(card))
+         {
+            Uninstall(card);
```

Comment:
`Uninstall` throws if the card was already not installed.  Are you sure it won't throw here?  Maybe make Uninstall non-throwing, and insert throws at other uses of it?

Is there a guarantee that `CancelInstall` aborts the response and that this callback is not invoked again?  I'm not sure.

Reply:
Removed those unnecessary checks from Uninstall. Now, it will only throw if there was an error in wxFileRemove when the file actually exists. CancelInstall should always abort the response